In [44]:
import pandas as pd 
import numpy as np

In [3]:
# Creating a numeric pandas series
s= pd.Series([2,4,5,6,8])
print(s)
print(type(s))

0    2
1    4
2    5
3    6
4    8
dtype: int64
<class 'pandas.core.series.Series'>


In [5]:
# creating a series of characters
# notice that the 'dtype' here is 'object'
char_series = pd.Series(['a', 'b', 'af'])
char_series

0     a
1     b
2    af
dtype: object

In [11]:
# creating a series of type datetime
date_series = pd.date_range(start = '09-11-2017', end = '12-11-2017')
print(date_series)
type(date_series)

DatetimeIndex(['2017-09-11', '2017-09-12', '2017-09-13', '2017-09-14',
               '2017-09-15', '2017-09-16', '2017-09-17', '2017-09-18',
               '2017-09-19', '2017-09-20', '2017-09-21', '2017-09-22',
               '2017-09-23', '2017-09-24', '2017-09-25', '2017-09-26',
               '2017-09-27', '2017-09-28', '2017-09-29', '2017-09-30',
               '2017-10-01', '2017-10-02', '2017-10-03', '2017-10-04',
               '2017-10-05', '2017-10-06', '2017-10-07', '2017-10-08',
               '2017-10-09', '2017-10-10', '2017-10-11', '2017-10-12',
               '2017-10-13', '2017-10-14', '2017-10-15', '2017-10-16',
               '2017-10-17', '2017-10-18', '2017-10-19', '2017-10-20',
               '2017-10-21', '2017-10-22', '2017-10-23', '2017-10-24',
               '2017-10-25', '2017-10-26', '2017-10-27', '2017-10-28',
               '2017-10-29', '2017-10-30', '2017-10-31', '2017-11-01',
               '2017-11-02', '2017-11-03', '2017-11-04', '2017-11-05',
      

pandas.core.indexes.datetimes.DatetimeIndex

In [14]:
# Indexing pandas series: Same as indexing 1-d numpy arrays or lists
# accessing the fourth element
s[3]
# accessing elements starting index = 2 till the end
s[2:]

# accessing the second and the fourth elements
# note that s[1, 3] will not work, you need to pass the indices [1, 3] as a list inside the original []
s[[1, 3]]

1    4
3    6
dtype: int64

#### Explicitly specifying indices
you might have noticed that while creating a series, Pandas automatically indexes it from 0 to (n-1), n being the number of rows. But if we want, we can also explicitly set the index ourselves, using the ‘index’ argument while creating the series using `pd.Series()`

In [15]:
# Indexing explicitly
pd.Series([0,1,2], index= ['a','b','c'])

a    0
b    1
c    2
dtype: int64

In [16]:
# You can also give the index as a sequence or use functions to specify the index
# But always make sure that the number of elements in the index list
pd.Series(np.array(range(0,10)) **2 ,index=range(0,10))

0     0
1     1
2     4
3     9
4    16
5    25
6    36
7    49
8    64
9    81
dtype: int64

### The Pandas Dataframe
Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having an index and columns having meaningful names.

In [17]:
# keys become column names
df = pd.DataFrame({'name': ['Vinay', 'Kushal', 'Aman', 'Saif'], 
                   'age': [22, 25, 24, 28], 
                    'occupation': ['engineer', 'doctor', 'data analyst', 'teacher']})
df

age    name    occupation
0   22   Vinay      engineer
1   25  Kushal        doctor
2   24    Aman  data analyst
3   28    Saif       teacher

#### Importing CSV data files as pandas dataframes 
For the upcoming exercises, we will use a dataset of a retail store having details about the orders placed, customers, product details, sales, profits etc. 


In [53]:
# reading a CSV file as a dataframe
market_df = pd.read_csv("/home/mahannee/Downloads/prep_DataAnalysis_docs/jupyter_notebook_docs/pandas/Introduction to Pandas/global_sales_data/market_fact.csv")


Usually, dataframes are imported as CSV files, but sometimes it is more convenient to convert dictionaries 
into dataframes. For e.g. when the raw data is in a JSON format (which is not uncommon), you can easily convert it into a dictionary, and then into a dataframe. 

You will learn how to convert JSON objects to dataframes later.

#### Reading and Summarising Dataframes
After you import a dataframe, you'd want to quickly understand its structure, shape, meanings of rows and columns etc. Further, you may want to look at summary statistics - such as mean, percentiles etc.

In [54]:
# Looking at the top and bottom entries of dataframes
market_df.head()
market_df.tail()

Ord_2265  Prod_id   Ship_id    Cust_id      Sales  Discount  \
8394  Ord_10660   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_10661   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_10662   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_10663  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_10664   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity  Profit  Shipping_Cost  Product_Base_Margin  
8394              28  374.63           7.69                 0.59  
8395              20  -74.03           6.92                 0.37  
8396              39  -70.85           5.35                 0.40  
8397              23  565.34          30.00                 0.62  
8398              47  131.39           4.86                 0.38

Here, each row represents an order placed at a retail store. Notice the index associated with each row - starts at 0 and ends at 8398, implying that there were 8399 orders placed.

In [55]:
# Looking at the datatypes of each column
market_df.info()
# Note that each column is basically a pandas Series of length 8399
# The ID columns are 'objects', i.e. they are being read as characters
# The rest are numeric (floats or int)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
Ord_2265               8399 non-null object
Prod_id                8399 non-null object
Ship_id                8399 non-null object
Cust_id                8399 non-null object
Sales                  8398 non-null float64
Discount               8399 non-null float64
Order_Quantity         8399 non-null int64
Profit                 8399 non-null float64
Shipping_Cost          8399 non-null float64
Product_Base_Margin    8336 non-null float64
dtypes: float64(5), int64(1), object(4)
memory usage: 656.2+ KB


In [24]:
# Describe gives you a summary of all the numeric columns in the dataset
market_df.describe()

Sales     Discount  Order_Quantity        Profit  Shipping_Cost  \
count   8399.000000  8399.000000     8399.000000   8399.000000    8399.000000   
mean    1775.878179     0.049671       25.571735    181.184424      12.838557   
std     3585.050525     0.031823       14.481071   1196.653371      17.264052   
min        2.240000     0.000000        1.000000 -14140.700000       0.490000   
25%      143.195000     0.020000       13.000000    -83.315000       3.300000   
50%      449.420000     0.050000       26.000000     -1.500000       6.070000   
75%     1709.320000     0.080000       38.000000    162.750000      13.990000   
max    89061.050000     0.250000       50.000000  27220.690000     164.730000   

       Product_Base_Margin  
count          8336.000000  
mean              0.512513  
std               0.135589  
min               0.350000  
25%               0.380000  
50%               0.520000  
75%               0.590000  
max               0.850000

In [25]:
# Column names
market_df.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [26]:
# The number of rows and columns
market_df.shape

(8399, 10)

In [27]:
# You can extract the values of a dataframe as a numpy array using df.values 
market_df.values

array([['Ord_5446', 'Prod_16', 'SHP_7609', ..., -30.51, 3.6, 0.56],
       ['Ord_5406', 'Prod_13', 'SHP_7549', ..., 4.56, 0.93, 0.54],
       ['Ord_5446', 'Prod_4', 'SHP_7610', ..., 1148.9, 2.5, 0.59],
       ..., 
       ['Ord_5388', 'Prod_6', 'SHP_7524', ..., -70.85, 5.35, 0.4],
       ['Ord_5348', 'Prod_15', 'SHP_7469', ..., 565.34, 30.0, 0.62],
       ['Ord_5459', 'Prod_6', 'SHP_7628', ..., 131.39, 4.86, 0.38]], dtype=object)

#### Indices 

An important concept in pandas dataframes is that of *row indices*. By default, each row is assigned indices starting from 0, and are represented at the left side of the dataframe. 

In [28]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

Now, arbitrary numeric indices are difficult to read and work with. Thus, you may want to change the indices of the df to something more meanigful.

Let's change the index to Ord_id (unique id of each order), so that you can select rows using the order ids directly.

In [30]:
# Setting index to Ord_id,
# if Inplace=True , then it will set permanently,
#if False it will set for temporary
market_df.set_index('Ord_id', inplace = False).head()


Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38

In [31]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

Having meaningful row labels as indices helps you to select (subset) dataframes easily. You will study selecting dataframes in the next section. 

#### Sorting dataframes
You can sort dataframes in two ways - 1) by the indices and 2) by the values.  


In [35]:
# Sorting by index
# axis = 0 indicates that you want to sort rows (use axis=1 for columns)
market_df.sort_index(axis = 0, ascending = False)

Ord_id  Prod_id   Ship_id    Cust_id       Sales  Discount  \
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798    603.6900      0.00   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798   3872.8700      0.03   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798    243.0500      0.02   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798    127.1600      0.10   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798   2841.4395      0.08   
8393  Ord_5348   Prod_8  SHP_7470  Cust_1798    356.7200      0.07   
8392  Ord_5384   Prod_9  SHP_7519  Cust_1798    181.5000      0.08   
8391  Ord_5435   Prod_4  SHP_7594  Cust_1798   1991.8985      0.07   
8390  Ord_5435  Prod_16  SHP_7594  Cust_1798    316.9900      0.04   
8389  Ord_4444  Prod_13  SHP_6192  Cust_1519    159.4100      0.00   
8388  Ord_4424   Prod_1  SHP_6165  Cust_1519    994.0400      0.03   
8387  Ord_2220   Prod_3  SHP_3019   Cust_851     41.0600      0.04   
8386  Ord_2324   Prod_7  SHP_3189   Cust_851    121.8700      0.07   
8385  Ord_1833   Prod_3  SHP_2527   Cust_637    611.1600      0.04   
8384  Ord_4620   Prod_3  SHP_6435  Cust_1577     59.6200      0.04   
8383  Ord_2722   Prod_1  SHP_3731  Cust_1006   3508.3300      0.04   
8382  Ord_2658   Prod_3  SHP_3636  Cust_1006    120.9800      0.00   
8381  Ord_2696   Prod_4  SHP_3691  Cust_1006   2836.0505      0.01   
8380  Ord_2772   Prod_6  SHP_3807  Cust_1006    308.9200      0.04   
8379  Ord_2722   Prod_5  SHP_3730  Cust_1006   1008.9500      0.04   
8378  Ord_2706   Prod_2  SHP_3705  Cust_1006   1361.9100      0.05   
8377  Ord_2722  Prod_12  SHP_3729  Cust_1006     34.0100      0.00   
8376  Ord_2624   Prod_8  SHP_3590  Cust_1006   1211.0000      0.00   
8375  Ord_2772   Prod_3  SHP_3806  Cust_1006   1413.8200      0.10   
8374  Ord_2658   Prod_5  SHP_3637  Cust_1006   1082.6600      0.08   
8373  Ord_2600  Prod_16  SHP_3560  Cust_1006    106.6400      0.10   
8372  Ord_2772   Prod_9  SHP_3806  Cust_1006     56.9000      0.03   
8371  Ord_2624   Prod_4  SHP_3591  Cust_1006   4924.1350      0.07   
8370  Ord_2696  Prod_13  SHP_3690  Cust_1006     62.7800      0.04   
8369  Ord_3633   Prod_3  SHP_5031  Cust_1274   1169.2600      0.02   
...        ...      ...       ...        ...         ...       ...   
29    Ord_2478  Prod_12  SHP_3395   Cust_931    139.9800      0.07   
28    Ord_2573   Prod_4  SHP_3527   Cust_931   3594.7435      0.05   
27    Ord_2405   Prod_9  SHP_3300   Cust_931   1062.6900      0.01   
26    Ord_2456   Prod_5  SHP_3367   Cust_931    109.5800      0.00   
25    Ord_2335  Prod_13  SHP_3204   Cust_931     38.2600      0.03   
24    Ord_2573   Prod_3  SHP_3525   Cust_931    197.6100      0.08   
23     Ord_996   Prod_7  SHP_1378   Cust_371     35.6400      0.05   
22     Ord_996   Prod_5  SHP_1377   Cust_371   3202.2500      0.09   
21     Ord_996   Prod_6  SHP_1378   Cust_371     81.2500      0.01   
20     Ord_996  Prod_13  SHP_1378   Cust_371     57.1700      0.08   
19     Ord_996  Prod_13  SHP_1378   Cust_371     41.9700      0.05   
18    Ord_4427   Prod_6  SHP_6171  Cust_1521    283.1300      0.08   
17    Ord_4471  Prod_15  SHP_6228  Cust_1521  13255.9300      0.02   
16    Ord_2282   Prod_9  SHP_3122   Cust_839    443.4600      0.06   
15    Ord_2280   Prod_5  SHP_3114   Cust_839    460.6900      0.06   
14    Ord_2207  Prod_10  SHP_3006   Cust_839   1410.9300      0.08   
13    Ord_2207  Prod_11  SHP_3093   Cust_839   3364.2480      0.10   
12    Ord_2978  Prod_16  SHP_4112  Cust_1088    305.0500      0.04   
11    Ord_1925   Prod_6  SHP_2637   Cust_708    465.9000      0.05   
10    Ord_4743   Prod_2  SHP_6615  Cust_1641   4072.0100      0.01   
9     Ord_4725   Prod_6  SHP_6593  Cust_1641     57.2200      0.07   
8     Ord_4725  Prod_13  SHP_6593  Cust_1641    162.0000      0.01   
7     Ord_4725   Prod_4  SHP_6593  Cust_1641   3410.1575      0.10   
6       Ord_31  Prod_12    SHP_41    Cust_26     14.7600      0.01   
5     Ord_5446   Prod_6  SHP_7608  Cust_1818    164.0200      0.03   
4     Ord_5485  Pr

In [39]:
# Sorting by values
# Sorting in increasing order of Sales
market_df.sort_values(by='Sales').head()

Ord_id  Prod_id   Ship_id    Cust_id     Sales  Discount  \
2738  Ord_3084  Prod_17  SHP_4279  Cust_1151  89061.05      0.00   
3784  Ord_2338  Prod_17  SHP_3207   Cust_932  45923.76      0.07   
1835  Ord_3875  Prod_17  SHP_5370  Cust_1351  41343.21      0.09   
2349  Ord_2373  Prod_14  SHP_3259   Cust_942  33367.85      0.01   
4399  Ord_4614  Prod_14  SHP_6423  Cust_1571  29884.60      0.05   

      Order_Quantity    Profit  Shipping_Cost  Product_Base_Margin  
2738              13  27220.69          24.49                 0.39  
3784               7    102.61          24.49                 0.39  
1835               8   3852.19          24.49                 0.39  
2349               9   3992.52          24.49                 0.37  
4399              49  12748.86          24.49                 0.44

In [40]:
# Sorting in decreasing order of Shipping_Cost
market_df.sort_values(by='Shipping_Cost', ascending = False).head()

Ord_id  Prod_id   Ship_id   Cust_id      Sales  Discount  \
4509  Ord_1751  Prod_15  SHP_2426  Cust_597  14740.510      0.00   
5327   Ord_839  Prod_11  SHP_1361  Cust_364  12689.870      0.04   
8283  Ord_1741  Prod_11  SHP_2411  Cust_595  15168.820      0.02   
2800   Ord_417  Prod_11   SHP_561  Cust_156  20333.816      0.02   
5511  Ord_1581  Prod_15  SHP_2184  Cust_519   2573.920      0.07   

      Order_Quantity   Profit  Shipping_Cost  Product_Base_Margin  
4509              46  3407.73         164.73                 0.56  
5327              44  -169.23         154.12                 0.76  
8283              26 -1096.78         147.12                 0.80  
2800              45 -1430.45         147.12                 0.80  
5511              17   117.23         143.71                 0.55

In [51]:
# Sorting by more than two columns
# Sorting in ascending order of Sales for each Product
market_df.sort_values(by=['Prod_id','Ord_2265' ,'Sales'], ascending = True).head(50)

Ord_2265 Prod_id   Ship_id    Cust_id     Sales  Discount  \
7743  Ord_10009  Prod_1  SHP_6996  Cust_1698    431.11      0.07   
7768  Ord_10034  Prod_1  SHP_4558  Cust_1183    550.29      0.03   
7772  Ord_10038  Prod_1  SHP_4452  Cust_1183    494.34      0.05   
7781  Ord_10047  Prod_1  SHP_4626  Cust_1196   1243.45      0.10   
7808  Ord_10074  Prod_1  SHP_4814  Cust_1188   1337.08      0.10   
7816  Ord_10082  Prod_1  SHP_4532  Cust_1188    329.37      0.04   
7855  Ord_10121  Prod_1  SHP_4538  Cust_1200    431.06      0.02   
7859  Ord_10125  Prod_1  SHP_4812  Cust_1251    876.01      0.07   
7903  Ord_10169  Prod_1  SHP_6085  Cust_1488   1225.41      0.04   
7909  Ord_10175  Prod_1  SHP_6986  Cust_1693   5875.66      0.02   
7921  Ord_10187  Prod_1  SHP_6332  Cust_1530    297.85      0.04   
7959  Ord_10225  Prod_1  SHP_5618  Cust_1376   2536.31      0.01   
7971  Ord_10237  Prod_1  SHP_2336   Cust_582    365.62      0.03   
7980  Ord_10246  Prod_1  SHP_3844  Cust_1035    317.82      0.02   
7982  Ord_10248  Prod_1  SHP_3739  Cust_1035   1199.57      0.06   
7984  Ord_10250  Prod_1  SHP_6441  Cust_1583   1264.13      0.02   
7991  Ord_10257  Prod_1  SHP_6299  Cust_1548    158.99      0.07   
8027  Ord_10293  Prod_1  SHP_4535  Cust_1219   1771.67      0.09   
8050  Ord_10316  Prod_1  SHP_1145   Cust_271   3401.80      0.09   
8055  Ord_10321  Prod_1  SHP_6637  Cust_1638    448.10      0.10   
8064  Ord_10330  Prod_1  SHP_7493  Cust_1805    522.62      0.02   
8067  Ord_10333  Prod_1  SHP_7655  Cust_1805     88.94      0.05   
8075  Ord_10341  Prod_1  SHP_7597  Cust_1795    539.82      0.00   
8101  Ord_10367  Prod_1  SHP_5495  Cust_1369    645.14      0.10   
8102  Ord_10368  Prod_1  SHP_5495  Cust_1369   1556.61      0.06   
8124  Ord_10390  Prod_1  SHP_5482  Cust_1363    294.91      0.03   
8125  Ord_10391  Prod_1  SHP_6285  Cust_1546   4976.60      0.06   
8143  Ord_10409  Prod_1  SHP_4530  Cust_1190  10330.94      0.07   
8149  Ord_10415  Prod_1  SHP_4723  Cust_1187    668.80      0.04   
8165  Ord_10431  Prod_1  SHP_2644   Cust_713    290.07      0.07   
8204  Ord_10470  Prod_1  SHP_4578  Cust_1180    312.25      0.05   
8205  Ord_10471  Prod_1  SHP_4632  Cust_1180    257.41      0.09   
8211  Ord_10477  Prod_1  SHP_2925   Cust_802   5318.89      0.06   
8230  Ord_10496  Prod_1  SHP_4658  Cust_1177   1011.67      0.04   
8237  Ord_10503  Prod_1  SHP_1066   Cust_245    770.53      0.04   
8268  Ord_10534  Prod_1  SHP_6447  Cust_1587    848.19      0.06   
8277  Ord_10543  Prod_1  SHP_6403  Cust_1522    234.24      0.09   
8280  Ord_10546  Prod_1  SHP_2423   Cust_595    679.52      0.09   
8286  Ord_10552  Prod_1  SHP_2424   Cust_595   5176.27      0.01   
8298  Ord_10564  Prod_1  SHP_2960   Cust_787   6648.58      0.08   
8323  Ord_10589  Prod_1  SHP_3819  Cust_1016   2014.13      0.10   
8349  Ord_10615  Prod_1  SHP_4905  Cust_1266   1184.11      0.07   
8383  Ord_10649  Prod_1  SHP_3731  Cust_1006   3508.33      0.04   
8388  Ord_10654  Prod_1  SHP_6165  Cust_1519    994.04      0.03   
44     Ord_2310  Prod_1  SHP_1634   Cust_452   1974.66      0.06   
57     Ord_2323  Prod_1  SHP_6327  Cust_1474   5208.78      0.05   
89     Ord_2355  Prod_1  SHP_7481  Cust_1802    123.85      0.09   
112    Ord_2378  Prod_1   SHP_286    Cust_45    127.80      0.07   
147    Ord_2413  Prod_1  SHP_1895   Cust_498   1538.17      0.02   
150    Ord_2416  Prod_1  SHP_2256   Cust_498    857.84      0.02   

      Order_Quantity   Profit  Shipping_Cost  Product_Base_Margin  
7743               4  -244.33          35.00                 0.82  
7768              30   -82.50           9.47                 0.59  
7772              15   -30.19           6.72                 0.75  
7781              39 -1212.55          35.00                 0.81  
7808              29  -953.00          35.00                 0.83  
7816              21   -53.57           4.53                 0.81  
7855              33   -26.78           6.13                 0

In [43]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38